client/core/trade.go

```diff
@@ -671,10 +625,316 @@ func (match *matchTracker) isActive() bool {
+		// taker's swap swapConf'th confirmations:
+		nextAction.maxWindow = bTimeout + // max time for maker's redeem to reach server
+			time.Minute + auth.DefaultConnectTimeout + // max time for server to validate maker's redeem and notify taker
+			time.Minute + bTimeout // max time for taker to validate maker's redeem and ack redemption request.
+
+	case status == order.MakerRedeemed && side == order.Taker:
+		// Server will wait a maximum of broadcastTimeout for taker's redeem,
+		// counting from the time the server acked maker's redemption. This
+		// time is communicated to taker in the server-originating redemption
+		// request.
+		nextAction.maxWindow = bTimeout
+
+	case status == order.MakerRedeemed && side == order.Maker:
+		// In additon to waiting the above-estimated max duration for taker's
+		// redeem, AND assuming the server does receive taker's redeem within
+		// that timeframe, maker can expect to wait the following additional
+		// durations before receiving redemption request from the server:
+		// - txWaitExpiration=1 minute; for server to find and validate taker's
+		//   redemption.
+		// - defaultConnectTimeout=10 minutes; for redemption request to reach
+		//   maker, assuming maker was temporarily disconnected.
+		//
+		// Also, the server will wait the following before revoking the match:
+		// - txWaitExpiration=1 minute; for maker to find and validate taker's
+		//   redemption.
+		// - broadcastTimeout for maker's ack of the redemption request.
+		//
+		// While the match may be revoked sooner, (e.g. if maker is disconnected
+		// for longer than 10 minutes and thus does not receive the redemption
+		// request), maker can only be certain of a revocation if the match remains
+		// at MakerRedeemed after the below max duration, counting from the time
+		// the server acked maker's redemption:
+		nextAction.maxWindow = bTimeout + // max time for taker's redeem to reach server
+			time.Minute + auth.DefaultConnectTimeout + // max time for server to validate taker's redeem and notify maker
+			time.Minute + bTimeout // max time for maker to validate taker's redeem and ack redemption request.
+	}
+
+	// Allow additional time for next action to prevent prematurely revoking this
+	// trade.
+	// TODO: Implement a match_status route to confirm match revocation with server
+	// before proceeding.
+	allowance := 10 * time.Minute
+	maxActionTime := nextAction.referenceTime.Add(nextAction.maxWindow).Add(allowance)
+	now := time.Now().Truncate(time.Millisecond).UTC()
+	if now.After(maxActionTime) {
+		log.Debugf("Auto-revoking match %v, status %s, side %s. Expected %s between %s and %s.",
+			match.id, status, side, nextAction.action, nextAction.referenceTime, maxActionTime)
+		t.setRevoked(match)
+	}
+}
+
+// checkSwapConfirmations checks for matches whose maker or taker swap have
+// received the reqiured confirmations and records the swap confirmation time.
+func (t *trackedTrade) checkSwapConfirmations() {
```

Comment:
Currently, this is only called on tip change as discussed in #629, and while that is preferable, it just occurred to me that trigger on tip change alone might not be sufficient.

If this `checkSwapConfirmations` method is not called soon after a new block is mined (e.g. if called after 2 new blocks), this client's next step on this match may be delayed beyond the permitted max action window, causing the match to be revoked.

While I'd prefer to not check swap confirmation count multiple times every `broadcastTimeout` (currently every 20 seconds given a default broadcastTimeout of 1 minute), it might be the safest thing to do. Thoughts?